File: src/main/java/net/easecation/ghosty/recording/level/updated/LevelUpdatedPopup.java
package net.easecation.ghosty.recording.level.updated;

import cn.nukkit.Player;
import cn.nukkit.utils.BinaryStream;
import net.easecation.ghosty.recording.level.LevelRecordNode;

public class LevelUpdatedPopup implements LevelUpdated {

    private String message;

    private LevelUpdatedPopup(String message) {
        this.message = message;
    }

    public LevelUpdatedPopup(BinaryStream stream) {
        this.read(stream);
    }

    public static LevelUpdatedPopup of(String message) {
        return new LevelUpdatedPopup(message);
    }

    @Override
    public int getUpdateTypeId() {
        return LevelUpdated.TYPE_POPUP;
    }

    @Override
    public void processTo(LevelRecordNode node) {
        node.offerLevelGlobalCallback((level) -> {
            for (Player player : level.getPlayers().values()) {
                player.sendPopup(this.message);
            }
        });
    }

    public void backwardTo(LevelRecordNode node) {
        // 不需要做任何事
    }

    @Override
    public void write(BinaryStream stream) {
        stream.putString(message);
    }

    @Override
    public void read(BinaryStream stream) {
        this.message = stream.getString();
    }

    @Override
    public boolean equals(Object obj) {
        if(!(obj instanceof LevelUpdatedPopup o)) return false;
        return this.message.equals(o.message);
    }

    @Override
    public String toString() {
        return "LevelUpdatedPopup{" +
            "message='" + message + '\'' +
            '}';
    }

    @Override
    public int hashCode() {
        return this.message.hashCode();
    }
}

File: src/main/java/net/easecation/ghosty/recording/level/updated/LevelUpdatedPopup.kt
package net.easecation.ghosty.recording.level.updated

import cn.nukkit.utils.BinaryStream
import kotlinx.serialization.Serializable
import net.easecation.ghosty.recording.level.LevelRecordNode

@Serializable
data class LevelUpdatedPopup(
    val message: String
) : LevelUpdated {

    override fun getUpdateTypeId(): Int = LevelUpdated.TYPE_POPUP

    override fun processTo(node: LevelRecordNode) {
        node.offerLevelGlobalCallback { level ->
            level.players.values.forEach { player ->
                player.sendPopup(message)
            }
        }
    }

    override fun backwardTo(node: LevelRecordNode) = Unit

    companion object {
        @JvmStatic
        fun of(message: String) = LevelUpdatedPopup(message)

        @JvmStatic
        val ADAPTER: LevelUpdateAdapter<LevelUpdatedPopup> = Adapter
    }

    private object Adapter : LevelUpdateAdapter<LevelUpdatedPopup> {
        override fun write(updated: LevelUpdatedPopup, stream: BinaryStream) {
            stream.putString(updated.message)
        }

        override fun read(stream: BinaryStream, formatVersion: Int) = LevelUpdatedPopup(
            message = stream.string
        )
    }
}